crates/notifier_host/src/watcher.rs

```diff
@@ -0,0 +1,258 @@
+use zbus::{dbus_interface, export::ordered_stream::OrderedStreamExt, Interface};
+
+pub const WATCHER_BUS_NAME: &'static str = "org.kde.StatusNotifierWatcher";
+pub const WATCHER_OBJECT_NAME: &'static str = "/StatusNotifierWatcher";
+
+async fn parse_service<'a>(
+    service: &'a str,
+    hdr: zbus::MessageHeader<'_>,
+    con: &zbus::Connection,
+) -> zbus::fdo::Result<(zbus::names::UniqueName<'static>, &'a str)> {
+    if service.starts_with("/") {
+        // they sent us just the object path :(
+        if let Some(sender) = hdr.sender()? {
+            Ok((sender.to_owned(), service))
+        } else {
+            log::warn!("unknown sender");
+            Err(zbus::fdo::Error::InvalidArgs("Unknown bus address".into()))
+        }
+    } else {
+        let busname: zbus::names::BusName = match service.try_into() {
+            Ok(x) => x,
+            Err(e) => {
+                log::warn!("received invalid bus name {:?}: {}", service, e);
+                return Err(zbus::fdo::Error::InvalidArgs(e.to_string()));
+            }
+        };
+
+        if let zbus::names::BusName::Unique(unique) = busname {
+            Ok((unique.to_owned(), "/StatusNotifierItem"))
+        } else {
+            // unwrap: we should always be able to access the dbus interface
+            let dbus = zbus::fdo::DBusProxy::new(&con).await.unwrap();
+            match dbus.get_name_owner(busname).await {
+                Ok(owner) => Ok((owner.into_inner(), "/StatusNotifierItem")),
+                Err(e) => {
+                    log::warn!("failed to get owner of {:?}: {}", service, e);
+                    Err(e)
+                }
+            }
+        }
+    }
+}
+
+/// Wait for a DBus service to exit
+async fn wait_for_service_exit(connection: zbus::Connection, service: zbus::names::BusName<'_>) -> zbus::fdo::Result<()> {
+    let dbus = zbus::fdo::DBusProxy::new(&connection).await?;
+    let mut owner_changes = dbus.receive_name_owner_changed_with_args(&[(0, &service)]).await?;
+
+    if !dbus.name_has_owner(service.as_ref()).await? {
+        return Ok(());
+    }
+
+    while let Some(sig) = owner_changes.next().await {
+        let args = sig.args()?;
+        if args.new_owner().is_none() {
+            break;
+        }
+    }
+
+    Ok(())
+}
+
+/// An instance of [`org.kde.StatusNotifierWatcher`].
+///
+/// [`org.kde.StatusNotifierWatcher`]: https://freedesktop.org/wiki/Specifications/StatusNotifierItem/StatusNotifierWatcher/
+#[derive(Debug, Default)]
+pub struct Watcher {
+    tasks: tokio::task::JoinSet<()>,
+    hosts: std::sync::Arc<std::sync::Mutex<std::collections::HashSet<String>>>,
+    items: std::sync::Arc<std::sync::Mutex<std::collections::HashSet<String>>>,
```

Comment:
Given that this is used in an async context, shouldn't we be using tokio::sync::Mutex instead?

Reply:
not really, see <https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use>. i'll leave a comment about it